Make positive -> finite with new facts like is_extended_positive

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Based on #16603 
Alternative to #16655 

#### Brief description of what is fixed or changed

This is an attempt to build on #16603 which makes is_real imply is_finite. The idea is to also make positive, nonpositive etc also imply finite so that positive implies real. In #16655 I tried to do this without introducing new facts and hit a dead end. This attempt introduces new facts such as is_extended_positive which allow for positive testing with oo and -oo included. The assumption rules used here are:
```python
_assume_rules = FactRules([

    'integer        ->  rational',
    'rational       ->  real',
    'rational       ->  algebraic',
    'algebraic      ->  complex & finite',
    'transcendental ==  complex & !algebraic & finite',
    'real           ->  hermitian',
    'imaginary      ->  complex & finite',
    'imaginary      ->  antihermitian',
    'extended_real  ->  commutative',
    'complex        ->  commutative',

    'odd            ==  integer & !even',
    'even           ==  integer & !odd',

    'real           ->  complex',
    'extended_real  ->  real | infinite',
    'real           ==  extended_real & finite',

    'extended_real        ==  extended_negative | zero | extended_positive',
    'extended_negative    ==  extended_nonpositive & extended_nonzero',
    'extended_positive    ==  extended_nonnegative & extended_nonzero',

    'extended_nonpositive ==  extended_real & !extended_positive',
    'extended_nonnegative ==  extended_real & !extended_negative',

    'real           ==  negative | zero | positive',
    'negative       ==  nonpositive & nonzero',
    'positive       ==  nonnegative & nonzero',

    'nonpositive    ==  real & !positive',
    'nonnegative    ==  real & !negative',

    'positive       ==  extended_positive & finite',
    'negative       ==  extended_negative & finite',
    'nonpositive    ==  extended_nonpositive & finite',
    'nonnegative    ==  extended_nonnegative & finite',
    'nonzero        ==  extended_nonzero & finite',

    'zero           ->  even & finite',
    'zero           ==  extended_nonnegative & extended_nonpositive',
    'zero           ==  nonnegative & nonpositive',

    'prime          ->  integer & positive',
    'composite      ->  integer & positive & !prime',
    '!composite     ->  !positive | !even | prime',

    'irrational     ==  real & !rational',

    'imaginary      ->  !extended_real',

    'infinite       ->  !finite',
    'noninteger     ==  extended_real & !integer',
    'extended_nonzero == extended_real & !zero',
])
```
Note that many of these rules are redundant. I initially had a more economical set of implications but deduce_all_facts was unable to make all possible inferences so I added additional rules to help it.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
    * The assumptions system is changed so that only finite numbers can be considered real, positive, negative, nonnegative, nonpositive or nonzero (since nonzero implies real). This means that any symbol declared with e.g. real=True is now automatically considered finite. It also means that infinities can not be considered positive or negative since they are not real (e.g. oo.is_positive is now False).
    * New assumptions extended_real, extended_positive etc are added that allow for positive and negative infinity. The equivalent of `Symbol('x', real=True)` in version 1.4 is now `Symbol('x', extended_real=True)`. The equivalent of `Symbol('x', negative=False)` is now `Symbol('x', extended_negative=False)` although it is usually better to use `Symbol('x', nonnegative=True)` (which implies both real=True and finite=True as well). Code that previously checked `if x.is_positive` should now be written as `if x.is_extended_positive` if it is intended that infinities should be allowed.
    * A number of corner cases in assumptions (e.g. for Pow.is_zero, Mul.is_positive etc) have been fixed to properly take account of the possibility of infinities.
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
Add extended_real and Make real -> finite

oscarbenjamin committed:
Replace is_real -> is_extended_real in numbers.py

oscarbenjamin committed:
Replace is_real -> is_extended_real in add.py

oscarbenjamin committed:
Replace is_real -> is_extended_real in expr.py

oscarbenjamin committed:
Replace is_real -> is_extended_real in basic.py

oscarbenjamin committed:
Replace is_real -> is_extended_real in mul.py

oscarbenjamin committed:
Replace is_real -> is_extended_real in power.py

oscarbenjamin committed:
Replace is_real -> is_extended_real in complexes.py

oscarbenjamin committed:
Replace is_real -> is_extended_real in trigonometric.py

oscarbenjamin committed:
Fix tests after making real imply finite

oscarbenjamin committed:
Add extended_real functions/elementary/miscellaneous.py

oscarbenjamin committed:
Use extended real for AccumBounds in calculus/util.py

oscarbenjamin committed:
WIP - ...

oscarbenjamin committed:
WIP - ...

oscarbenjamin committed:
WIP ...

oscarbenjamin committed:
WIP ...

oscarbenjamin committed:
WIP ...

oscarbenjamin committed:
WIP...

oscarbenjamin committed:
is_real -> is_extended_real in functions/elementary.py

oscarbenjamin committed:
Fix tests for is_real in functions/elementary.py

oscarbenjamin committed:
is_real -> is_extended_real in functions/special.py

oscarbenjamin committed:
Fix tests for is_real in function/special

oscarbenjamin committed:
is_real -> is_extended_real in simplify

oscarbenjamin committed:
is_real -> is_extended_real in inequalities

oscarbenjamin committed:
is_real -> is_extended_real in solvers.py

oscarbenjamin committed:
extended_real should not imply complex

oscarbenjamin committed:
WIP ...

oscarbenjamin committed:
Make extended_real -> commutative

oscarbenjamin committed:
Fix test real->extended_real

oscarbenjamin committed:
Add _eval_is_complex to re/im

oscarbenjamin committed:
Make imaginary -> finite

oscarbenjamin committed:
Fix tests for is_extended_real

oscarbenjamin committed:
is_real -> is_extended_real in exprtools.py

oscarbenjamin committed:
is_real -> is_extended_real in exponential.py

oscarbenjamin committed:
Fix test for is_real -> finite

oscarbenjamin committed:
Add is_finite to LambertW

oscarbenjamin committed:
Use finite=True in LambertW test

oscarbenjamin committed:
Fix test fail for is_real -> finite in complexes

oscarbenjamin committed:
Fix test for real -> finite

oscarbenjamin committed:
Fix test for is_real -> finite in logic

oscarbenjamin committed:
Fix test for is_real -> is_finite in quantum

oscarbenjamin committed:
Fix fail from is_real -> finite in series

oscarbenjamin committed:
Fix test for is_extended_real in simplify

oscarbenjamin committed:
is_real -> is_extended_real in integrals

oscarbenjamin committed:
Fix test in solveset for real -> finite

oscarbenjamin committed:
Fix wester test for real -> finite

oscarbenjamin committed:
Fix stats and tests for real -> finite

oscarbenjamin committed:
Improve assumptions for Pow

oscarbenjamin committed:
Fix transforms and tests for real -> finite

oscarbenjamin committed:
Update test in solvers for extended_real assumption

oscarbenjamin committed:
Update tests in powsimp for new assumptions

oscarbenjamin committed:
real->extended_real in sets tests

oscarbenjamin committed:
Make CRootOf finite

oscarbenjamin committed:
Fix assumptions in test_inequalities for real->finite

oscarbenjamin committed:
Fix assumptions in test_rationaltools

oscarbenjamin committed:
Fix assumption in test_manual for real->finite

oscarbenjamin committed:
Add _eval_is_real to error functions

oscarbenjamin committed:
Fix doctests for real -> finite

oscarbenjamin committed:
Use extended_real=True in solveset tests

oscarbenjamin committed:
Use extended_real=True in test_integrals

oscarbenjamin committed:
Update homogeneous_order for real -> finite

oscarbenjamin committed:
Restore Wester test test_Y2

oscarbenjamin committed:
Add erf._eval_is_finite

oscarbenjamin committed:
Work around hang in Weibull using finite=True...

oscarbenjamin committed:
Fix tests in integrals for real -> finite

oscarbenjamin committed:
XFAILed two failing tests in series

oscarbenjamin committed:
Fix tests and comment hanging assert in meijerint

oscarbenjamin committed:
Update test for real->finite in test_ode.py

oscarbenjamin committed:
XFAIL test_W23 in test_wester due to real->finite

oscarbenjamin committed:
Fix AccumBounds.is_real and remove XFAIL

oscarbenjamin committed:
Fix and uncomment failing test

oscarbenjamin committed:
Remove XFAIL from test_order.py

oscarbenjamin committed:
Use is_extended_real in polyroots and remove XFAIL

oscarbenjamin committed:
Remove redundant assumption rule in ask.py

oscarbenjamin committed:
Remove redundant infinity checks

oscarbenjamin committed:
Remove redundant finite checks

oscarbenjamin committed:
Add extended_real in output for test_tree

oscarbenjamin committed:
WIP

oscarbenjamin committed:
WIP making positive -> finite in core...

oscarbenjamin committed:
positive -> extended_positive in Add

oscarbenjamin committed:
positive -> extended_positive in Expr

oscarbenjamin committed:
Fix core tests for positive -> finite

oscarbenjamin committed:
positive -> extended_positive in Abs

oscarbenjamin committed:
positive -> extended_positive in Pow

oscarbenjamin committed:
Add _eval_is handlers in trigonometric

oscarbenjamin committed:
positive -> extended_positive in exponential

oscarbenjamin committed:
Fix tests for positive -> finite in exponential

oscarbenjamin committed:
positive -> extended_positive in periodic_argument

oscarbenjamin committed:
positive -> extended_positive in Pow.as_real_imag

oscarbenjamin committed:
Comment failing test in gamma_functions

oscarbenjamin committed:
Update new assumptions tests for positive->finite

oscarbenjamin committed:
positive -> extended_positive in AccumBounds

oscarbenjamin committed:
positive -> extended_positive in Infinity._eval_power

oscarbenjamin committed:
positive->extended_positive in SingularityFunction

oscarbenjamin committed:
Update tests assumptions tests for extended_positive etc

oscarbenjamin committed:
Update test for extended_positive etc assumptions

oscarbenjamin committed:
positive=False -> nonpositive=True in stats tests

oscarbenjamin committed:
positive -> extended_positive in sign

oscarbenjamin committed:
positive -> extended_positive in _pt

oscarbenjamin committed:
XFAIL 3 tests failing due to positive -> finite

oscarbenjamin committed:
positive -> extended_positive in function

oscarbenjamin committed:
Fix doctests in basic/printing for positive -> finite

oscarbenjamin committed:
Comment out doctest...

jksuom commented on sympy/calculus/util.py:
```diff
@@ -995,17 +996,17 @@ class AccumulationBounds(AtomicExpr):
-        if not (min.is_real or min in inftys) \
-           or not (max.is_real or max in inftys):
+        # Only allow real intervals (use symbols with 'is_extended_real=True').
+        if not (min.is_extended_real or min in inftys) \
```
Is `min in inftys` redundant? `oo.is_extended_real` should be true.

oscarbenjamin replied:
Yes it is redundant

oscarbenjamin replied:
I've removed the infinity check

jksuom commented on sympy/core/add.py:
```diff
@@ -493,8 +491,8 @@ def _eval_is_algebraic_expr(self, syms):
         return all(term._eval_is_algebraic_expr(syms) for term in self.args)
 
     # assumption methods
-    _eval_is_real = lambda self: _fuzzy_group(
```
Should this also be be preserved?

oscarbenjamin replied:
I think it should. I'll add it back...

oscarbenjamin replied:
I've added it back

oscarbenjamin committed:
Re-add _eval_is_positive to Expr

oscarbenjamin commented:
I've just pushed a commit to fix the failing test. The issue was due to Expr not having _eval_is_positive.

I think that maybe the code for Expr._eval_is_extended_positive/negative should be factored out. If I understand correctly the code currently in _eval_is_extended_positive doesn't handle infinities and should really be in a _eval_is_positive method. Expr needs to define _eval_is_extended_positive so that it can be called from Add._eval_is_extended_positive in
```
     return super(Add, self)._eval_is_extended_positive()
```
I'm not sure if that can just be changed back to calling _eval_is_positive because I don't know if super would always find Expr as the superclass and other classes might implement _eval_is_extended_positive differently taking account of infinities as well.

The commit that I just pushed includes (in Expr):
```python
     def _eval_is_extended_positive(self):
        if self.is_extended_real is False:
            return False
        if Expr._eval_is_positive(self):
            return True
```

oscarbenjamin commented:
This PR currently has 3 XFAILS.

There are two for limits:
```console
$ pytest sympy/utilities/tests/test_wester.py --runxfail -k test_T6
============================================================= test session starts =============================================================
...
================================================================== FAILURES ===================================================================
___________________________________________________________________ test_T6 ___________________________________________________________________

    @XFAIL
    def test_T6():
>       assert limit(1/n * factorial(n)**(1/n), n, oo) == exp(-1)
E       assert 0 == exp(-1)
E        +  where 0 = limit(((1 / n) * (factorial(n) ** (1 / n))), n, oo)
E        +    where factorial(n) = factorial(n)
E        +  and   exp(-1) = exp(-1)

sympy/utilities/tests/test_wester.py:2093: AssertionError
                                                               DO *NOT* COMMIT!                                                               
================================================== 1 failed, 396 deselected in 1.03 seconds ===================================================
```

```console
$ pytest sympy/series/tests/test_limits.py --runxfail -k test_factorial
============================================================= test session starts =============================================================
================================================================== FAILURES ===================================================================
_______________________________________________________________ test_factorial ________________________________________________________________

    @XFAIL
    def test_factorial():
        from sympy import factorial, E
        f = factorial(x)
        assert limit(f, x, oo) == oo
>       assert limit(x/f, x, oo) == 0
E       AssertionError: assert Limit(x/factorial(x), x, oo, dir='-') == 0
E        +  where Limit(x/factorial(x), x, oo, dir='-') = limit((x / factorial(x)), x, oo)

sympy/series/tests/test_limits.py:412: AssertionError
                                                               DO *NOT* COMMIT!                                                               
=================================================== 1 failed, 61 deselected in 0.70 seconds ===================================================
```

The third XFAIL is to do with the new assumptions calling the old assumptions. I think that's because I haven't updated the new assumptions yet but I'm not sure:
```console
$ pytest sympy/assumptions/tests/test_sathandlers.py --runxfail -k test_CheckOldAssump
============================================================= test session starts =============================================================
================================================================== FAILURES ===================================================================
_____________________________________________________________ test_CheckOldAssump _____________________________________________________________

    @XFAIL
    def test_CheckOldAssump():
        # TODO: Make these tests more complete
    
        class Test1(Expr):
            def _eval_is_positive(self):
                return True
            def _eval_is_negative(self):
                return False
    
        class Test2(Expr):
            def _eval_is_finite(self):
                return True
            def _eval_is_positive(self):
                return True
            def _eval_is_negative(self):
                return False
    
        t1 = Test1()
        t2 = Test2()
    
        # We can't say if it's positive or negative in the old assumptions without
        # bounded. Remember, True means "no new knowledge", and
        # Q.positive(t2) means "t2 is positive."
>       assert CheckOldAssump(Q.positive(t1)) == True
E       assert Q.positive(Test1()) == True
E        +  where Q.positive(Test1()) = CheckOldAssump(Q.positive(Test1()))
E        +    where Q.positive(Test1()) = Q.positive(Test1())
E        +      where Q.positive = <sympy.assumptions.ask.AssumptionKeys object at 0x108696320>.positive

sympy/assumptions/tests/test_sathandlers.py:94: AssertionError
                                                               DO *NOT* COMMIT!                                                               
=================================================== 1 failed, 5 deselected in 0.18 seconds ====================================================
```

oscarbenjamin commented on sympy/concrete/products.py:
```diff
@@ -96,8 +96,9 @@ class Product(ExprWithIntLimits):
-    pi/2
+
+    #>>> limit(W2e, n, oo)
+    #pi/2
```
I have commented out this doctest because it failed:
```
092     >>> from sympy import limit
093     >>> W2 = Product(2*i/(2*i-1)*2*i/(2*i+1), (i, 1, n))
094     >>> W2
095     Product(4*i**2/((2*i - 1)*(2*i + 1)), (i, 1, n))
096     >>> W2e = W2.doit()
097     >>> W2e
098     2**(-2*n)*4**n*factorial(n)**2/(RisingFactorial(1/2, n)*RisingFactorial(3/2, n))
099 
100     >>> limit(W2e, n, oo)
Expected:
    pi/2
Got:
    Limit(2**(-2*n)*4**n*factorial(n)**2/(RisingFactorial(1/2, n)*RisingFactorial(3/2, n)), n, oo, dir='-')

sympy/concrete/products.py:100: DocTestFailure
```

oscarbenjamin committed:
Remove redundant infinity check in AccumBounds

oscarbenjamin committed:
Re-add Add._eval_is_real

oscarbenjamin committed:
Re-add Pow._eval_is_positive

oscarbenjamin commented:
The tests have passed (with 3 XFAILED and one doctest commented out) but have taken longer than usual to run. I can reproduce a slowdown with:
```
$ bin/test sympy/simplify/tests/test_hyperexpand.py -k test_formulae --slow
```
That takes 24 seconds with this PR and 14 seconds on master.

oscarbenjamin committed:
Add _eval_is_positive back to Expr

oscarbenjamin committed:
Add _eval_is_positive to Float

oscarbenjamin committed:
real -> extended_real in Expr.is_extended_positive

oscarbenjamin committed:
Fix assumptions for atan

oscarbenjamin committed:
Reimplement Expr._eval_is_positive

oscarbenjamin commented:
There is an intermittent test failure that I can reproduce on Python 3.7 with:
```console
$ PYTHONHASHSEED=2937395614 bin/test sympy/core/tests/test_assumptions.py -k test_Mul_is_prime_composite --seed=73309503
============================================================= test process starts =============================================================
executable:         /Users/enojb/current/sympy/venv/bin/python  (3.7.1-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       gmpy 1.17
numpy:              1.16.1
random seed:        73309503
hash randomization: on (PYTHONHASHSEED=2937395614)

sympy/core/tests/test_assumptions.py[1] F                                                                                                [FAIL]

_______________________________________________________________________________________________________________________________________________
______________________________________ sympy/core/tests/test_assumptions.py:test_Mul_is_prime_composite _______________________________________
Traceback (most recent call last):
  File "/Users/enojb/current/sympy/sympy/sympy/core/tests/test_assumptions.py", line 834, in test_Mul_is_prime_composite
    assert ( (x+1)*(y+1) ).is_prime is False
AssertionError

============================================= tests finished: 0 passed, 1 failed, in 0.28 seconds =============================================
DO *NOT* COMMIT!
```

oscarbenjamin committed:
Fix recursion problems in Expr._eval_is_composite

oscarbenjamin committed:
Simplify Expr._eval_is_positive

oscarbenjamin committed:
Add _eval_is_finite to re and im

oscarbenjamin committed:
use extended_positive and un-XFAIL tests in series

oscarbenjamin commented:
I think I've fixed the performance problem. I've also fixed the 2 XFAILed tests and one commented test that were all to do with limits.

That leaves one XFAILed test to do with the new assumptions:
```console
$ pytest sympy/assumptions/tests/test_sathandlers.py --runxfail -k test_CheckOldAssump
```
I think that's just because I should update the new assumptions now.

oscarbenjamin commented:
It looks like the new assumptions already define positive as finite:
https://github.com/sympy/sympy/blob/3335542a72800e1c982af941ea5dd58ad73ce007/sympy/assumptions/sathandlers.py#L186

So I think the only problem there is that the failing test is no longer valid under the changes in this PR:
```console
$ pytest sympy/assumptions/tests/test_sathandlers.py --runxfail -k test_CheckOldAssump
============================================================= test session starts =============================================================
================================================================== FAILURES ===================================================================
_____________________________________________________________ test_CheckOldAssump _____________________________________________________________

    @XFAIL
    def test_CheckOldAssump():
        # TODO: Make these tests more complete
    
        class Test1(Expr):
            def _eval_is_positive(self):
                return True
            def _eval_is_negative(self):
                return False
    
        class Test2(Expr):
            def _eval_is_finite(self):
                return True
            def _eval_is_positive(self):
                return True
            def _eval_is_negative(self):
                return False
    
        t1 = Test1()
        t2 = Test2()
    
        # We can't say if it's positive or negative in the old assumptions without
        # bounded. Remember, True means "no new knowledge", and
        # Q.positive(t2) means "t2 is positive."
>       assert CheckOldAssump(Q.positive(t1)) == True
E       assert Q.positive(Test1()) == True
E        +  where Q.positive(Test1()) = CheckOldAssump(Q.positive(Test1()))
E        +    where Q.positive(Test1()) = Q.positive(Test1())
E        +      where Q.positive = <sympy.assumptions.ask.AssumptionKeys object at 0x108696320>.positive

sympy/assumptions/tests/test_sathandlers.py:94: AssertionError
```

oscarbenjamin committed:
UnXFAIL test with positive -> extended_positive

oscarbenjamin commented:
I've fixed that test with positive -> extended_positive. That means there are no failing tests with this PR now.

Todo:
1. Go back through the diff
2. Audit uses of is_positive, is_real throughout codebase
3. Investigate performance (tests are running maybe 5% slower right now)

Reviews/suggestions welcome!

oscarbenjamin committed:
Restore complex -> finite | infinite

oscarbenjamin commented:
I've rebased on #16603 which was rebased onto master because of #16592

oscarbenjamin committed:
positive -> extended_positive in Infinity.__mul__

oscarbenjamin commented:
Rebasing to include #16727

oscarbenjamin commented on sympy/core/numbers.py:
```diff
@@ -920,7 +920,7 @@ class Float(Number):
     is_number = True
 
-    is_real = True
+    is_extended_real = True
```
@smichr  should this be is_real=True as well after #16727?

jksuom replied:
It has not yet been decided whether Infinity should derive from Float or not (see https://github.com/sympy/sympy/pull/16727#issuecomment-487182859),

smichr replied:
@asmeurer says,

> In short, perhaps we can leave this as is for now, until a need for oo to be a Float becomes clear. It never happened with NaN, and maybe it will never happen with oo either.

And I suggested that extended real be `{-oo, Float, oo}` (implying that oo not be a Float). That 'inf' is represented by an mpf is not an argument that it should be a Float -- mpfs are "oo" aware just as SymPy is aware of it (as a Singleton).

I would vote to keep it as it is currently: `oo.is_Float -> False`.

It would make sense to tweak the Float instantiation to reject `nan`, `inf` and `-inf` in favor of instantiating those with `Number` instead since there is nothing to be gained by allowing them to instantiate from Float -- Float just returns the Singleton (which does not have the `_mpf_` attribute of Floats but can be coaxed to tell it as in `oo._as_mpf_val(1) -> (0, 0, -456, -2)`.

smichr replied:
@asmeurer also said:

>  I don't know if having oo be a true Float with an _mpf_ would make internal code cleaner, such as the code in evalf.py.

And I don't know whether that would make things better or not. But even still, we might consider a class `Real(Float)` or `ExtendedReal(Float)` which simply allows `_mpf_` to be requested directly instead of requiring special handling of `oo` with `_as_mpf_val`.

jksuom replied:
I think that such classes might make the handling of real numbers even more confusing. There is already the `RealElement` class that is not derived from `Float` although both have attributes of `mpf` type.

jksuom replied:
As to the question of `Float.is_real`, I think that it should be True. It seems that others concur.

oscarbenjamin replied:
Thanks. I've re-added is_real=True so that Float.is_real is always True.

oscarbenjamin commented on sympy/core/tests/test_arit.py:
```diff
@@ -378,14 +378,15 @@ def test_Add_Mul_is_integer():
 
     assert sin(x).is_finite is True
-    assert (x*sin(x)).is_finite is False
+    assert (x*sin(x)).is_finite is None
```
Should `(x*sin(x)).is_finite` be False when x.is_finite is False?

I think of this as a nan type situation:
```
In [9]: (x*sin(x)).subs(x, oo)                                                                                                    
Out[9]: <-∞, ∞>

In [10]: print((x*sin(x)).subs(x, oo).is_zero)                                                                                    
None

In [11]: print((x*sin(x)).subs(x, oo).is_finite)                                                                                  
None
```

oscarbenjamin committed:
Remove unneeded changes

oscarbenjamin committed:
Update test_arit for positive -> finite

oscarbenjamin committed:
Add nonzero -> real

oscarbenjamin committed:
Add tests for extended_positive

oscarbenjamin committed:
Add test for real=True in test_expr

oscarbenjamin committed:
nonnegative -> extended_nonnegative in power.py

oscarbenjamin committed:
Assumptions cleanup for positive->finite in complexes

oscarbenjamin committed:
Fix assumptions for atan

oscarbenjamin committed:
Test cleanup for positive -> finite

oscarbenjamin commented on sympy/functions/special/tests/test_gamma_functions.py:
```diff
@@ -65,8 +65,10 @@ def test_gamma():
+    # exp_polar.is_integer but I'm not sure if that makes sense.
     assert gamma(3*exp_polar(I*pi)/4).is_nonnegative is False
-    assert gamma(3*exp_polar(I*pi)/4).is_nonpositive is True
+    assert gamma(3*exp_polar(I*pi)/4).is_extended_nonpositive is True
```
Does it make sense to define exp_polar.is_integer? If so should it exp_polar(x).is_integer be the same as exp_polar(x)?

jksuom replied:
It is not easy to say what that would mean. Technically, exp_polar is not even complex (except possibly for real arguments).

oscarbenjamin replied:
Does it even make sense to use exp_polar as an argument to the gamma function?

oscarbenjamin commented on sympy/integrals/tests/test_integrals.py:
```diff
@@ -593,7 +593,7 @@ def test_integrate_max_min():
+    c = symbols('c', extended_real=True)
     int1 = integrate(Max(c, x)*exp(-x**2), (x, -oo, oo))
     int2 = integrate(c*exp(-x**2), (x, -oo, c))
     int3 = integrate(x*exp(-x**2), (x, c, oo))
```
This test gives different answers depending on whether c is declared as real or extended_real. Here I've just changed it to extended_real to match previous behaviour but maybe something needs fixing in integrals. The difference for int3 is:
```julia
In [16]: c = symbols('c', real=True)                                                                                              

In [17]: integrate(x*exp(-x**2), (x, c, oo))                                                                                      
Out[17]: 
  ⎛                     2⎞               
  ⎜                   -c ⎟               
  ⎜                2⋅ℯ   ⎟               
c⋅⎜-2⋅√π⋅erfc(c) + ──────⎟               
  ⎝                  c   ⎠   √π⋅c⋅erfc(c)
────────────────────────── + ────────────
            4                     2      

In [18]: c = symbols('c', extended_real=True)                                                                                     

In [19]: integrate(x*exp(-x**2), (x, c, oo))                                                                                      
Out[19]: 
   2
 -c 
ℯ   
────
 2  
```
For int2 the difference is:
```julia
In [20]: c = symbols('c', real=True)                                                                                              

In [21]: integrate(c*exp(-x**2), (x, -oo, c))                                                                                     
Out[21]: 
√π⋅c⋅(2 - erfc(c))
──────────────────
        2         

In [22]: c = symbols('c', extended_real=True)                                                                                     

In [23]: integrate(c*exp(-x**2), (x, -oo, c))                                                                                     
Out[23]: 
√π⋅c⋅erf(c)   √π⋅c
─────────── + ────
     2         2  
```

smichr replied:
hmm the negative sign is missing.

oscarbenjamin replied:
I think the result above for int2 with extended_real=True is correct. It is also the same as the result on master with real=True. The result for real=True on this PR is incorrect:
```julia
In [1]: c = symbols('c', real=True)                                                                                               

In [2]: integrate(c*exp(-x**2), (x, -oo, c))                                                                                      
Out[2]: 
√π⋅c⋅(2 - erfc(c))
──────────────────
        2 
```
That's a separate issue from this PR though because I get the same on master if I declare c to be finite:
```julia
In [1]: c = symbols('c', real=True, finite=True)                                                                                  

In [2]: integrate(c*exp(-x**2), (x, -oo, c))                                                                                      
Out[2]: 
√π⋅c⋅(2 - erfc(c))
──────────────────
        2
```

oscarbenjamin committed:
positive->extended_positive and cleanup in special

oscarbenjamin committed:
Cleanup for positive->finite in integrals

oscarbenjamin committed:
Remove unneeded test line

oscarbenjamin committed:
Remove redundant finite=True from tests

oscarbenjamin commented on sympy/solvers/inequalities.py:
```diff
@@ -204,7 +204,7 @@ def reduce_rational_inequalities(exprs, gen, relational=True):
     >>> from sympy.solvers.inequalities import reduce_rational_inequalities
 
-    >>> x = Symbol('x', real=True)
+    >>> x = Symbol('x', extended_real=True)
```
On master this doctest gives:
```julia
In [1]: from sympy.solvers.inequalities import reduce_rational_inequalities                                                       

In [2]: x = Symbol('x', real=True)                                                                                                

In [3]: reduce_rational_inequalities([[x + 2 > 0]], x)                                                                            
Out[3]: -2 < x ∧ x < ∞
```
However this PR gives:
```
In [3]: reduce_rational_inequalities([[x + 2 > 0]], x)                                                                            
Out[3]: -2 < x
```
This is because of x is real then x>oo auto-evaluates to true. I've changed it to extended_real=True so that the test output stays the same.

Would it be better to change the output for these tests (there are also examples)?

smichr replied:
or just include both tests so it is clear that we know they are different and are ok with that

oscarbenjamin replied:
We can but I don't really know how to explain this to a user: why when x is declared as extended_real (and therefore possibly infinite) does the solution come back explicitly excluding oo?

oscarbenjamin replied:
I've changed it to real=True and included an example with extended_real=True.

oscarbenjamin committed:
Remove unneeded finite=True in ode

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -2918,7 +2918,7 @@ def nonlinsolve(system, *symbols):
     >>> from sympy import pprint
     >>> from sympy.polys.polytools import is_zero_dimensional
-    >>> a, b, c, d = symbols('a, b, c, d', real=True)
+    >>> a, b, c, d = symbols('a, b, c, d', extended_real=True)
```
This doctest is also in the main tests and is
```python
from sympy import pprint, symbols, nonlinsolve
from sympy.polys.polytools import is_zero_dimensional
a, b, c, d = symbols('a, b, c, d', real=True)
eq1 =  a + b + c + d
eq2 = a*b + b*c + c*d + d*a
eq3 = a*b*c + b*c*d + c*d*a + d*a*b
eq4 = a*b*c*d - 1
system = [eq1, eq2, eq3, eq4]
is_zero_dimensional(system)
pprint(nonlinsolve(system, [a, b, c, d]), use_unicode=False)
```
On master this gives:
```
  -1       1               1      -1              
{(---, -d, -, {d} \ {0}), (-, -d, ---, {d} \ {0})}
   d       d               d       d           
```
However under this PR it is
```
  -1       1       1      -1      
{(---, -d, -, d), (-, -d, ---, d)}
   d       d       d       d      
```
I've changed it to extended_real to preserve the original output. I'm not sure which is actually better though.

Is the test intentionally demonstrating something with the exclusion of zero? Should I instead change the test output?

smichr replied:
Something is wrong, I think, because being extended doesn't change the fact that `d` cannot be 0.

oscarbenjamin replied:
So eq4 cannot be satisfied if d=0 unless a, b or c is infinite. That would seem pertinent but the difference in behaviour is the wrong way round. When a, b, and c are declared as real (i.e. finite under this PR) it then becomes possible for d to be zero...

oscarbenjamin replied:
The same difference occurs on master if the symbols are declared with finite=True. Perhaps a separate issue should be opened for this.

oscarbenjamin replied:
I've opened #16909

oscarbenjamin committed:
Clean up and remove unnecessary finite=True in solvers tests

oscarbenjamin committed:
Remove unneeded finite=True from stats

oscarbenjamin commented:
I've just been through the full diff made some changes and cleaned some things up. I left a couple of questions above.

The slowdown I saw before seems to have solved itself looking at the timing of the last Travis run.

I haven't done a full audit of the code base for places where real=True, is_positive etc. are used but e.g. the output of `git grep finite=True` shows that most uses of finite=True can be removed after the changes here.

It has taken a lot of work to get to this point and the size of this patch means that it rapidly becomes stale as new PRs get merged so the longer this is open the more conflicts I'll need to fix. If this is going to go in then I think most further changes should be in separate PRs.

I'm happy with this now so this is ready for review.

oscarbenjamin committed:
Make Float.is_real always True

ShubhamKJha commented on sympy/functions/special/error_functions.py:
```diff
@@ -1409,7 +1415,7 @@ def fdiff(self, argindex=1):
         z = self.args[0]
         # Exclude values on the branch cut (-oo, 0)
-        if not (z.is_real and z.is_negative):
+        if not z.is_negative:
```
Shouldn't this be `z.is_extended_negative` ?

oscarbenjamin replied:
Yes, I've changed it to that

ShubhamKJha commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -338,7 +338,7 @@ def test_bessel_nan():
 def test_conjugate():
     from sympy import conjugate, I, Symbol
-    n, z, x = Symbol('n'), Symbol('z', real=False), Symbol('x', real=True)
+    n, z, x = Symbol('n'), Symbol('z', extended_real=False), Symbol('x', real=True)
```
I believe it should be `n, z, x = Symbol('n'), Symbol('z', extended_real=False), Symbol('x', extended_real=True)`

oscarbenjamin replied:
In this PR BesselBase._eval_conjugate is
```
    def _eval_conjugate(self):
        z = self.argument
        if z.is_extended_negative is False:
            return self.__class__(self.order.conjugate(), z.conjugate())
```
I changed the test to exclude the possibility of z being -oo but now I wonder if I `_eval_conjugate` should be changed instead...

oscarbenjamin replied:
Actually I think that `_eval_conjugate` is fine. I've changed it to `Symbol('x', extended_real=True)` as suggested.

ShubhamKJha reviewed (commented):
@oscarbenjamin we can use a shorter notation for `_is_extended_property` like `is_Positive` for `is_extended_positive`.

asmeurer commented on sympy/core/mul.py:
```diff
@@ -1159,19 +1164,19 @@ def _eval_is_polar(self):
 
         for t in self.args:
-            if t.is_complex is False:
+            if t.is_complex is False and t.is_extended_real is False:
```
I thought complex implied finite. extended_real could include oo or -oo.

oscarbenjamin replied:
On master and in this PR complex does not imply finite and zoo is considered complex in both cases.

(It probably wouldn't be so hard to build on this PR to make complex -> finite -- actually that was my intention but I wanted to get this merged before further work. Thanks for reviewing!)

As @asmeurer says the additional check is needed to exclude oo and -oo. On master we have
```
In [9]: oo.is_complex
Out[9]: True
```
whereas in this PR that would be False. Here oo and -oo are extended_real but not real. Since real implies complex oo is not complex.

oscarbenjamin replied:
Actually that doesn't quite follow as necessary but oo and -oo are deliberately not complex in this PR in any case.

ShubhamKJha replied:
We can iterate over this PR to make `complex -> finite` throughout. But then, `zoo` would not be `complex` anymore.

jksuom replied:
Maybe we should have `extended_complex` for `zoo`.

asmeurer commented:
> @oscarbenjamin we can use a shorter notation for _is_extended_property like is_Positive for is_extended_positive.

`is_Capitalized` already has a different meaning in SymPy (member of the `Capitalized` class). I personally like the `extended` names, as they are explicit and hopefully quite clear what they mean. 

I don't recall if we decided what `x > 0` should mean, finite positive or extended positive. If it means extended positive, then that could be a shorthand inside the new assumptions.

smichr commented:
> I don't recall if we decided what `x > 0` should mean, finite positive or extended positive.

Hopefullt this means extended since we wouldn't want `oo > 0` to fail, would we? `I > 0` is not the same as `oo > 0`.

smichr committed:
Merge branch 'master' into ext_pos_finite

smichr commented:
How about `is_(xreal, xpositive, etc...)` for the shorter form? or `is_real_oo`, `positive_oo` which explicitly shows that you are including `oo`

oscarbenjamin commented on sympy/core/add.py:
```diff
@@ -591,40 +591,40 @@ def _eval_is_irrational(self):
                 return
         return False
 
-    def _eval_is_positive(self):
```
Add can use _eval_is_positive from Expr. I added new _eval_is_positive/negative methods there

oscarbenjamin commented on sympy/core/mul.py:
```diff
@@ -1353,20 +1358,21 @@ def _eval_is_even(self):
             return False
 
     def _eval_is_composite(self):
-        if self.is_integer and self.is_positive:
```
This check is redundant because we check that each arg is an integer below. It might seem like the check above is a shortcut but it also loops through all args checking that they are integers.

In general it is important that an _eval_is_y method doesn't use self.is_x (it should check properties on args not self) although there are exceptions. It becomes problematic when self.is_x comes back to self.is_y because we have what would be a recursion error except for this line:
https://github.com/sympy/sympy/blob/master/sympy/core/assumptions.py#L283
Although that line stops recursion errors it leads to other weirdness like getting a different value for the same property when accessing it multiple times. The fix is to ensure that _eval_is methods only query properties of args rather than self.

oscarbenjamin commented on sympy/core/numbers.py:
```diff
@@ -2806,12 +2819,14 @@ def __le__(self, other):
-        if other.is_real:
-            if other.is_finite or other is S.NegativeInfinity:
+        if other.is_extended_real:
+            if other is S.Infinity:
```
I'll need to take a closer look at the logic here...

oscarbenjamin replied:
I've simplified the logic here now

oscarbenjamin commented on sympy/core/power.py:
```diff
@@ -453,42 +467,44 @@ def _eval_is_positive(self):
-            if self.exp.is_odd:
+    def _eval_is_extended_negative(self):
+        if self.base.is_extended_negative:
+            if self.exp.is_odd and self.base.is_finite:
```
The exponent is odd but the base can still be infinite. This is for e.g. `oo**-1`

ShubhamKJha replied:
Current master fails for this case:
```py
In [33]: p = symbols('p', negative=True)

In [34]: q = symbols('q', odd=True)

In [35]: (p**q).is_negative 
Out[35]: True # Shoud be None since p can be q can be negative as well
```

asmeurer replied:
This is actually a good example of why it's good to make positive -> finite. Without that, we cannot make a useful deduction here, because the possibility of infinity breaks the fact. Of course, it's a restricted assumption, so anything that is true of extended_positive is also true of positive, but not the other way around. So the end result is fewer `None`s from infinity technicalities. Fewer bugs too, as this shows, because people forget to consider infinity all the time.

oscarbenjamin replied:
Before this PR in #16433 I tried to fix the bugs in Pow.is_zero etc without changing the definition of any of the assumptions. If you look there you can see how many parts of the code base I had to change because they were depending on bugs in the assumptions system.

oscarbenjamin commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -102,15 +102,22 @@ def test_infinity():
-    assert oo.is_complex is True
+    assert oo.is_extended_real is True
+    assert oo.is_real is False
+    assert oo.is_complex is False
```
I don't think extended_complex is needed in the same way. extended_real is all about ordering and having oo and -oo and extreme elements is important. The same doesn't really apply to complex which is also why I think making complex imply finite would be easy relative to this.

oscarbenjamin replied:
Making complex imply finite (and adding extended_complex) can certainly be a separate PR

oscarbenjamin commented:
Thanks for the review @ShubhamKJha. I'll need to find some time to go through it.

@asmeurer @smichr 
Here x>y is equivalent to (x-y).is_extended_positive. That's how `Expr.__gt__` is ultimately implemented.

is_xreal or is_real_oo are both fine by me. I also don't think it's a problem for the name to be long though. Ideally it wouldn't be used as much as is_real etc anyway.

smichr commented:
> is_xreal or is_real_oo are both fine by me

Although "extended real" shows up as a standard way of referring to the concept, the `oo` form might be more mnemonic in nature for those that don't use this often. But that can be a point of education, on the other hand. I don't have a strong feeling either way and defer to those with greater mathematical sensibilities.

oscarbenjamin committed:
Add real=True in stats tests

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into ext_pos_finite

oscarbenjamin committed:
Add real=True in stats tests

oscarbenjamin committed:
Simplify logic in Infinity.__le__/__gt__

oscarbenjamin commented on sympy/core/numbers.py:
```diff
@@ -2820,12 +2835,14 @@ def __gt__(self, other):
-        if other.is_real:
-            if other.is_finite or other is S.NegativeInfinity:
+        if other.is_extended_real:
+            if other is S.Infinity:
```
I've simplified this now

oscarbenjamin commented on sympy/core/power.py:
```diff
@@ -510,30 +526,30 @@ def _eval_is_integer(self):
             check = self.func(*self.args)
             return check.is_Integer
 
-    def _eval_is_real(self):
```
It isn't strictly needed since is_real can be inferred from `_eval_is_extended_real` and `_eval_is_finite`. If added back it would be very similar to the `_eval_is_extended_real` method. Note also that adding an `_eval_is` method can sometimes speed things up and sometimes slow things down so this needs to be considered carefully for a core class like Pow.

The `Pow._eval_is_` methods can certainly be cleaned up and no doubt still have bugs. I think that can come in another PR though.

oscarbenjamin committed:
Simplify Pow._eval_is_positive

oscarbenjamin commented on sympy/core/power.py:
```diff
@@ -424,23 +424,37 @@ def _eval_is_even(self):
+        ext_neg = Pow._eval_is_extended_negative(self)
+        if ext_neg is True:
+            return self.is_finite
+        elif ext_neg is False:
```
Done

oscarbenjamin commented on sympy/core/power.py:
```diff
@@ -424,23 +424,37 @@ def _eval_is_even(self):
+        ext_neg = Pow._eval_is_extended_positive(self)
+        if ext_neg is True:
+            return self.is_finite
+        elif ext_neg is False:
```
Done

oscarbenjamin commented on sympy/core/tests/test_arit.py:
```diff
@@ -999,28 +1000,28 @@ def test_Pow_is_real():
 
-    x = Symbol("x", nonnegative=True)
-    y = Symbol("y", nonnegative=True)
+    x = Symbol("x", nonnegative=True, finite=True)
```
Agreed. I guess that's a hangover from #16603. I've reverted it.

oscarbenjamin commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -413,31 +456,31 @@ def test_symbol_falsepositive():
     # To test pull request 9379
     # Explicit handling of arg.is_positive=False was added to Mul._eval_is_positive
-    x = 2*Symbol('x', positive=False)
+    x = 2*Symbol('x', positive=False, real=True)
```
Yes this test was failing. The problem is this:
```
In [1]: x = Symbol('x', positive=False)                                                                                           

In [2]: x.is_positive                                                                                                             
Out[2]: False

In [3]: (2*x).is_positive                                                                                                         

In [4]: print((2*x).is_positive)                                                                                                  
None
```
I guess that can be fixed by adding Mul._eval_is_positive

oscarbenjamin replied:
Looks like master has the same problem with nonpositive:
```
In [1]: x = Symbol('x', nonpositive=False)                                                                                        

In [2]: x.is_nonpositive                                                                                                          
Out[2]: False

In [3]: (2*x).is_nonpositive                                                                                                      

In [4]: print((2*x).is_nonpositive)                                                                                               
None
```

ShubhamKJha replied:
This is happening because `mul` object doesn't seem to have it's own `_eval_is_nonpositive` or `_eval_is_nonnegative` methods and completely depends on the `assumptions` provided by `ManagedProperties` which doesn't imply anything.
```py
>>> from sympy import symbols
>>> x = symbols('x', nonpositive=False)
>>> (2*x).is_nonpositive
>>> (2*x)._assumptions
{'nonpositive': None, 'imaginary': None, 'zero': False, 'antihermitian': None, 'complex':
 None, 'integer': None, 'rational': None, 'irrational': None, 'even': None, 'odd': None,
'finite': None, 'composite': None, 'infinite': None, 'algebraic': None, 'commutative': Tr
ue, 'negative': False, 'positive': None, 'hermitian': None, 'real': None}
>>> (2*x).assumptions0
{}
>>> x.assumptions0
{'nonpositive': False, 'negative': False, 'zero': False, 'commutative': True}
>>>
```
Needs these classes to be added.

oscarbenjamin replied:
The Mul `_eval_is` methods would need to be reorganised to fix it for nonpositive=False. I think I've fixed the issue for positive=False though so that I can restore these tests. Just running tests locally before pushing.

oscarbenjamin replied:
I've just pushed the change to restore these tests

oscarbenjamin committed:
Remove unnedded finite=True

oscarbenjamin commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -916,6 +916,7 @@ def _eval_template_is_attr(self, is_attr):
     _eval_is_odd = lambda self: self._eval_template_is_attr('is_odd')
     _eval_is_polar = lambda self: self._eval_template_is_attr('is_polar')
     _eval_is_positive = lambda self: self._eval_template_is_attr('is_positive')
+    _eval_is_extended_real = lambda self: self._eval_template_is_attr('is_extended_real')
```
Yes, I've added them

oscarbenjamin committed:
Add Piecewise._eval_is_extended_positive etc

oscarbenjamin committed:
Use is_extended_negative in erf

ShubhamKJha commented on sympy/core/numbers.py:
```diff
@@ -2956,31 +2956,21 @@ def __le__(self, other):
-                return S.true
+        if other.is_infinite and other.is_extended_positive:
+            return S.true
+        elif other.is_real or other.is_extended_nonpositive:
```
This can be further reduced to `other.is_extended_real`.

oscarbenjamin replied:
This is `oo <= other`. If I have `other = Symbol('x', extended_real=True)` then other could possibly be oo and `oo <= oo` evaluates to True.

ShubhamKJha replied:
That case is handled in the `if` part. I am suggesting for the `elif` part.

smichr replied:
I get:
```
is_real -> a=(-oo,oo)
extended_nonpositive -> b=[-oo, 1)
U(a,b)->[-oo,oo)
```
but `extended_real -> [-oo,oo]`, which is not the same. I think the result is 
```
other is not S.Infinity and other.is_extended_real
```

oscarbenjamin replied:
If other is oo then it will be handled in the if. If other is `Symbol('x', extended_real=True)` it should pass through both cases and the result should be unevaluated because `x` *might* be oo but it also might not:
```
>>> from sympy import *
>>> x = Symbol('x', extended_real=True)
>>> oo <= x
oo <= x
>>> oo <= oo
True
```

oscarbenjamin replied:
We can't use `other is not S.Infinity` because it won't exclude a symbol with the same assumptions i.e. `Symbol('y', extended_positive=True, infinite=True)` or a symbol that is *possibly* infinite e.g. `Symbol('x', extended_real=True)`.

oscarbenjamin committed:
Fix Mul._eval_is_positive for positive=False

oscarbenjamin commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -413,31 +456,31 @@ def test_symbol_falsepositive():
 
 def test_neg_symbol_falsepositive():
-    x = -Symbol('x', positive=False)
+    x = -Symbol('x', positive=False, real=True)
```
I've restored this

oscarbenjamin committed:
Use extended_real=True in bessel test

oscarbenjamin committed:
Update reduce_rational_inequalities docstring

oscarbenjamin commented:
Thanks you very much for taking the time to review this @ShubhamKJha. I think I've responded to all of your comments now but please let me know if I've missed something.

I think I've done everything needed in this PR now apart from squashing the commits. Is there any useful way to format them?

I don't think that the changes can be separated into independently revertible parts but I can split it into maybe one commit for the changes to the core, another for functions, and several for different parts of the tests.

smichr commented:
I'm not sure that keeping all the commit separate will help in tracking down any future bugs, but one will be able to use bisection to still tell if it is before or after this batch that something broke so I don't think there is a need to squash them.

This needs a release note telling what has changed, i.e. what worked before vs what works now. Is this backwards incompatible in assumptions in any way?

oscarbenjamin commented:
I've added a release note.

This is backwards incompatible. I find it hard to gauge how many problems users will have with it. I had to change a lot of things from e.g. is_real to is_extended_real but I suspect that this is more a problem in the core than it would be for users.

Perhaps a more likely problem for users and one that came up a few times in this PR is when assumptions are declared negatively e.g. `Symbol('x', negative=False)`. Under this PR that includes `-oo` as a possibility so on master we have:
```
In [1]: x = Symbol('x', negative=False)                                                                                           

In [2]: x < 0                                                                                                                     
Out[2]: False
```
Where as with this PR:
```
In [2]: x < 0                                                                                                                     
Out[2]: x < 0
```
I suspect that most users (and contributors?) who write negative=False don't realise what it actually means and should really use nonnegative=True or include real=True. I've mentioned this in the release notes.

smichr commented:
@jksuom , @asmeurer - any objections to this being committed?

asmeurer commented:
I have no objections but let's wait for a sign-off from @ShubhamKJha.

asmeurer commented:
What is the status of the new assumptions with this?

oscarbenjamin commented:
I don't understand what the situation with the new assumptions is. There is Q.real and also Q.extended_real but Q.real does not imply finite and Q.real implies Q.extended_real:
```
In [70]: print(ask(Q.finite(x), Q.real(x)))                                                                                       
None

In [2]: ask(Q.extended_real(x), Q.real(x))                                                                                        
Out[2]: True
```
Also this code suggests that the new assumptions are different somehow in this respect but I don't see how in terms of output:
https://github.com/sympy/sympy/blob/3335542a72800e1c982af941ea5dd58ad73ce007/sympy/assumptions/sathandlers.py#L186

oscarbenjamin commented:
Whoops! Q.real implies Q.extended_real is not surprising. That's what this PR does with the old assumptions.

In the new assumptions Q.extended_real does not imply Q.real:
```
In [6]: print(ask(Q.real(x), Q.extended_real(x)))                                                                                 
None
```
However Q.real does not imply finite:
```
In [7]: print(ask(Q.finite(x), Q.real(x)))                                                                                        
None
```
so I don't know what that means.

asmeurer commented:
We can clean up the new assumptions in a new PR. The two should ultimately be consistent with one another. I just wanted to see if that was already the case or not.

ShubhamKJha commented:
> In the new assumptions Q.extended_real does not imply Q.real:

That is correct. `extended_real` can also be `infinite`.

ShubhamKJha commented:
Thanks @oscarbenjamin this PR looks good.

There are certainly two issues that popped up:
* Inconsistent results with `nonpositive = False`
* Inconsistency in new assumptions

These can be taken care of in another PR.

jksuom commented:
I think it is time to merge this.

smichr commented:
> more likely problem for users and one that came up a few times in this PR is when assumptions are declared negatively

I think you have the right behavior here `negative=False` doesn't imply anything about being real since `I.is_negative` is also False.

oscarbenjamin commented:
Thanks everyone for reviewing and for merging.

There are a number of followup tasks that I'll list here:

1. Update the docs
2. Make the new assumptions consistent with the old
3. Make complex imply finite

For 3. I don't have time for a while now but I can get to it at some point. If someone else wants to do it sooner then go ahead! I didn't update the docs yet because I viewed this as a step towards making complex imply finite. The docs will need updating before the next release in any case though.

Also it looks like the core `_eval_is` methods for Pow, Add and Mul have room for improvement and there are a number of missing methods. #16433 will be somewhat invalidated by this merge but you can see there that just focussing closely on Pow.is_zero turned into a substantial patch. So that makes another task:

4. Closely review `_eval_is` methods for Pow, Add and Mul and add missing methods.

oscarbenjamin commented:
> > more likely problem for users and one that came up a few times in this PR is when assumptions are declared negatively
> 
> I think you have the right behavior here `negative=False` doesn't imply anything about being real since `I.is_negative` is also False.

It's the right behaviour for the wrong reason because after this PR we still have:
```
In [1]: x = Symbol('x', extended_negative=False)                                                                                  

In [2]: x < 0                                                                                                                     
Out[2]: False
```

This is due to this code in `Expr.__lt__` (that I didn't substantively change):
https://github.com/sympy/sympy/blob/b9604c5f59c587c8ecc5461e62ae70d96a4bfda5/sympy/core/expr.py#L365-L369

I have to say I don't understand exactly why that code is written the way it is but it doesn't handle the negative assumption case correctly.
